core/src/shred.rs

```diff
@@ -646,4 +925,210 @@ mod tests {
                 .verify(keypair.pubkey().as_ref(), &shred[coding_data_offset..]));
         }
     }
+
+    #[test]
+    fn test_recovery_and_reassembly() {
+        let keypair = Arc::new(Keypair::new());
+        let slot = 0x123456789abcdef0;
+        let mut shredder =
+            Shredder::new(slot, Some(5), 1.0, &keypair, 0).expect("Failed in creating shredder");
+
+        assert!(shredder.shreds.is_empty());
+        assert_eq!(shredder.active_shred, None);
+        assert_eq!(shredder.active_offset, 0);
+
+        let data: Vec<_> = (0..5000).collect();
+        let data: Vec<u8> = data.iter().map(|x| *x as u8).collect();
+        let mut offset = shredder.write(&data).unwrap();
+        offset += shredder.write(&data[offset..]).unwrap();
+        offset += shredder.write(&data[offset..]).unwrap();
+        offset += shredder.write(&data[offset..]).unwrap();
+        offset += shredder.write(&data[offset..]).unwrap();
+
+        // We should have 4 shreds now
+        assert_eq!(shredder.shreds.len(), 4);
+        assert_eq!(offset, data.len());
+
+        shredder.finalize_fec_block();
+
+        // We should have 10 shreds now (one additional final shred, and 5 coding shreds)
+        assert_eq!(shredder.shreds.len(), 10);
```

Comment:
instead of hard-coding these, how about talking about why there should be 4, becaus 5000/MTU, etc...